### ISSUE TITLE
Deployment Issue with pac4j-saml client initialization

### ISSUE THREAD

**Issue by lemurdance** [pac4j/pac4j#353]
I have a working SAML implementation using pac4j-saml and j2e-pac4j on Tomcat, and it runs well in staging, but on a production deployment the initialization is throwing an exception chain that's founded in the following ClassCastException when Pac4J loads up: 

Caused by: java.lang.ClassCastException: org.opensaml.xml.signature.impl.CryptoBinaryBuilder cannot be cast to org.opensaml.core.xml.XMLObjectBuilder
        at org.opensaml.core.xml.config.XMLConfigurator.initializeObjectProviders(XMLConfigurator.java:238)
        at org.opensaml.core.xml.config.XMLConfigurator.load(XMLConfigurator.java:203)
        at org.opensaml.core.xml.config.XMLConfigurator.load(XMLConfigurator.java:188)
        at org.opensaml.core.xml.config.XMLConfigurator.load(XMLConfigurator.java:162)
        at org.opensaml.core.xml.config.AbstractXMLObjectProviderInitializer.init(AbstractXMLObjectProviderInitializer.java:52)
        at org.opensaml.core.config.InitializationService.initialize(InitializationService.java:56)
        at org.pac4j.saml.util.Configuration.bootstrap(Configuration.java:76)

More to the point, the production-built WAR also runs fine on the staging device, so it's not the compiled and deployed WAR at issue. 

Searching for other similar problems online, it appears that there's often a conflict of two different OpenSAML packages on the same system, but I don't believe we're having that problem, since it works in one environment, and the production environment is a fresh plain-vanilla Tomcat container with every build. The underlying OS doesn't seem to have any RPMs installed that involve SAML. 

Since it's OpenSAML throwing the exception, but the source is Pac4J, I was hoping you could help me understand what server/container factors might cause this change, if you've seen it before with Pac4J deployments, and what I might do to solve the issue. 

I'm in a logjam with it, and don't know what the conflict might be at this point. 

Thanks very much!

**Comment by leleuj** [comment]
Hi,

Even with the same WAR, loading can happen differently between
environements: is is the same OS? The same JDK? The same Tomcat?

I find CryptoBInaryBuilder in org.opensaml / xmltooling v1.3.4:
http://grepcode.com/file/repo1.maven.org/maven2/org.opensaml/xmltooling/1.3.4/org/opensaml/xml/signature/impl/CryptoBinaryBuilder.java
while XMLObjectBuilder is in org.opensaml / opensaml-core v3.1.0:
http://grepcode.com/file/repo1.maven.org/maven2/org.opensaml/opensaml-core/3.1.0/org/opensaml/core/xml/XMLObjectBuilder.java

Can you try removing the xmltooling jar?

Thanks.
Best regards,
Jérôme


2015-11-13 21:06 GMT+01:00 Kevin Elliott <notifications@github.com>:

**Comment by leleuj** [comment]
Please use the mailing list for any questions.

**Comment by lemurdance** [comment]
Hi Jérôme,

Thanks so much, and removing the tooling jar (1.4.4 by dependency) breaks, as it's definitely necessary. I did try forcing use of 1.3.4, which works fine in the environments that are already good, but wasn't able to test, as the production environment itself (a corporate entity with stringent requirements) doesn't seem to have accepted the forced downgrade, so I can't test readily at the moment. In all cases, I appreciate your quick response and thought, and it's among the more bizarre issues I've run across. 

I'm turning it over to the production IT group to try and resolve (by adjusting the environment), as I have three working instances with a wide variety of variables shifted around without causing the specific failure in question. Sorry to bother you, and if and as we discover what the issue was, I'll keep you apprised. Very happy with how your work integrated easily and we have working SAML while staying lightweight. 

Thanks again!

**Comment by leleuj** [comment]
You're welcome. Keep us posted on the mailing list: it might help others.

**Comment by jkacer** [comment]
Hello!

May I ask if you resolved this issue somehow? Unfortunately I was not able to find the discussion on the mailing list.

This issue suddenly appeared yesterday on one of our test environments after we had migrated from Java 7 + Tomcat 7 to Java 8 + Tomcat 8.

I studied this discussion: https://issues.shibboleth.net/jira/browse/OSJ-152

As I understand it, the issue is caused by multiple initializers using the same configuration resource on the class path: default-config.xml. They assumed that each initializer would load its own config file but that's not the case - they both load the same file and the pick-up order is probably non-deterministic or it depends on the version of JRE/Tomcat (not sure).

Would it be possible to remove XML Tooling from the dependencies?
It is used at 2 points in 2 classes in PAC4J-SAML (class XmlHelper is used there). Maybe the same functionality could be implemented directly in PAC4J; I don't know how complex it is, though.
Or, alternatively, we could extract a small subset of XML Tooling (without the initializer etc) and provide it as a dependency for PAC4J-SAML instead.
Or, we could just fork XML Tooling and just rename the configuration file to a unique name. There would be no non-determinism then.

What do you think?
I am referring to PAC4J 1.7, maybe it has been solved since then, so please correct me if the above statements are not true anymore.

Thank you! Jarda

**Comment by leleuj** [comment]
I don't remember when, but it was done. Regarding the dependencies of pac4j 1.9.2-SNAPSHOT, I don't see it anymore.

```
[INFO] +- org.pac4j:pac4j-saml:jar:1.9.2-SNAPSHOT:compile
[INFO] |  +- org.opensaml:opensaml-core:jar:3.1.1:compile
[INFO] |  +- net.shibboleth.utilities:java-support:jar:7.1.1:compile
[INFO] |  |  \- com.google.code.findbugs:jsr305:jar:3.0.0:compile
[INFO] |  +- org.opensaml:opensaml-saml-api:jar:3.1.1:compile
[INFO] |  |  \- org.opensaml:opensaml-storage-api:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-saml-impl:jar:3.1.1:compile
[INFO] |  |  \- org.opensaml:opensaml-soap-impl:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-soap-api:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-xmlsec-api:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-security-api:jar:3.1.1:compile
[INFO] |  |  \- org.apache.santuario:xmlsec:jar:2.0.3:compile
[INFO] |  |     \- org.codehaus.woodstox:woodstox-core-asl:jar:4.4.1:compile
[INFO] |  |        +- javax.xml.stream:stax-api:jar:1.0-2:compile
[INFO] |  |        \- org.codehaus.woodstox:stax2-api:jar:3.1.4:compile
[INFO] |  +- org.opensaml:opensaml-security-impl:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-profile-api:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-profile-impl:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-messaging-api:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-messaging-impl:jar:3.1.1:compile
[INFO] |  +- org.opensaml:opensaml-xmlsec-impl:jar:3.1.1:compile
[INFO] |  +- com.google.guava:guava:jar:19.0:compile
[INFO] |  +- org.cryptacular:cryptacular:jar:1.1.0:compile
[INFO] |  +- joda-time:joda-time:jar:2.9.2:compile
[INFO] |  +- xalan:xalan:jar:2.7.2:compile
[INFO] |  |  \- xalan:serializer:jar:2.7.2:compile
[INFO] |  +- org.apache.velocity:velocity:jar:1.7:compile
[INFO] |  |  +- commons-collections:commons-collections:jar:3.2.1:compile
[INFO] |  |  \- commons-lang:commons-lang:jar:2.4:compile
[INFO] |  \- net.shibboleth.ext:spring-extensions:jar:5.1.1:compile
[INFO] |     \- org.apache.httpcomponents:httpclient-cache:jar:4.3.6:compile
```

Though, pac4j v1.7 is an old version: can you try upgrading to more recet versions?

**Comment by jkacer** [comment]
Thank you very much, Jerome! We will either upgrade or look at the change and try to incorporate the change.

**Comment by jkacer** [comment]
I can confirm that the new code from PAC4J 1.9 solves the issue :-)

BTW, the issue has not appeared until now because Tomcat 8 changes the way how JARs in WEB-INF/lib are loaded. Before, they were always sorted alphabetically, now they are not, so it depends only on the underlying file systems, how the JARs will be ordered. Moreover, it only appeared on Linux, not on Windows. Obviously their file systems report those files in different order.

**Comment by jhegg** [comment]
@jkacer, I apologize that I should have followed up on the bug that I opened on https://issues.shibboleth.net/jira/browse/OSJ-152, sorry about that. As @leleuj said and you have found out, the defect has been fixed as of pac4j v1.8.5, please see #444. I'm glad that you were able to resolve it quickly, and I'll make sure my defect over on the Shibboleth project is amended to have this information.

**Comment by jkacer** [comment]
Hi @jhegg and thank you for this additional information, I was not aware of this issue before.

What surprises me is that XML Tooling 1.4 is a part of Open SAML v. 2 (as those guys from Open SAML say), that was not obvious to me at all. I also found that 1.4.4 was the last released version, so I did not believe it was a part of Open SAML.

Gentlemen, thank you once again for your replies!
